### ISSUE TITLE
test/conversions.sh is failing on TZ test

### ISSUE THREAD

**Issue by clanmills** [Exiv2/exiv2#485]
This is occurs on MinGW and MSVC

**MSVC**
```
C:\Users\rmills\gnu\github\exiv2\buildserver\build>msys64                                                                                                                                                       
157 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/build $ cd ../test                                                                                                                                     
158 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $ PATH="/c/Python36:$PATH"                                                                                                                        
159 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $ env EXIV2_BINDIR=${PWD}/../build/bin ./conversions.sh                                                                                           
Files /c/Users/rmills/gnu/github/exiv2/buildserver/test/tmp/conversions.out and /c/Users/rmills/gnu/github/exiv2/buildserver/test/data/conversions.out differ                                                   
165c165                                                                                                                                                                                                         
< Exif.Image.DateTime                          Ascii      20  2015:04:17 19:10:22                                                                                                                               
---                                                                                                                                                                                                             
> Exif.Image.DateTime                          Ascii      20  2015:04:18 02:10:22                                                                                                                               
160 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $  
```

Here's a way to produce this from the terminal (without using test/conversions.sh) on an MSVC build:

```
C:\Users\rmills\gnu\github\exiv2\buildserver\test>msys64                                                                                                                                                        
162 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $ cd ~/gnu/github/exiv2/buildserver/test                                                                                                          
163 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $ cp data/exiv2-empty.jpg empty.jpg                                                                                                               
164 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $ ../build/bin/exiv2 -eX --verbose --force empty.jpg                                                                                              
File 1/1: empty.jpg                                                                                                                                                                                             
165 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $ ../build/bin/exiv2 -M'set Xmp.xmp.ModifyDate 2015-04-17T18:10:22Z' empty.xmp                                                                    
166 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $  TZ=GMT-8 ../build/bin/exiv2 -iX --verbose --force empty.jpg                                                                  
File 1/1: empty.jpg                                                                                                                                                                                             
Writing Exif data from .\empty.xmp to empty.jpg                                                                                                                                                                 
Writing XMP data from .\empty.xmp to empty.jpg                                                                                                                                                                  
168 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $ ../build/bin/exiv2 -pa --grep date/i empty.jpg                                                                                                  
Exif.Image.DateTime                          Ascii      20  2015:04:17 19:10:22                                                                                                                                 
Xmp.xmp.ModifyDate                           XmpText    20  2015-04-17T18:10:22Z                                                                                                                                
169 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/buildserver/test $ 
```

On MacOS-X, the same commands:
```
540 rmills@rmillsmm:~/gnu/github/exiv2/buildserver $ cd test
541 rmills@rmillsmm:~/gnu/github/exiv2/buildserver/test $ cp data/exiv2-empty.jpg empty.jpg
542 rmills@rmillsmm:~/gnu/github/exiv2/buildserver/test $ ../build/bin/exiv2 -eX --verbose --force empty.jpg
File 1/1: empty.jpg
543 rmills@rmillsmm:~/gnu/github/exiv2/buildserver/test $ ../build/bin/exiv2 -M'set Xmp.xmp.ModifyDate 2015-04-17T18:10:22Z' empty.xmp
544 rmills@rmillsmm:~/gnu/github/exiv2/buildserver/test $ TZ=GMT-8 ../build/bin/exiv2 -iX --verbose --force empty.jpg
File 1/1: empty.jpg
Writing Exif data from ./empty.xmp to empty.jpg
Writing XMP data from ./empty.xmp to empty.jpg
546 rmills@rmillsmm:~/gnu/github/exiv2/buildserver/test $ ../build/bin/exiv2 -pa --grep date/i empty.jpg 
Exif.Image.DateTime                          Ascii      20  2015:04:18 02:10:22
Xmp.xmp.ModifyDate                           XmpText    20  2015-04-17T18:10:22Z
547 rmills@rmillsmm:~/gnu/github/exiv2/buildserver/test $ 
```

**Analysis**
The command `TZ=GNT-8 exiv2 -iX ....`
1) is correctly copying xmp:ModifyDate to the XMP
2) the convertor on MSVC sets Exif.Image.DateTime to 2015:04:17 19:10:22 and it should be 2015:04:18 02:10:22 on my machine in England (during Daylight saving time).

MSVC:
```
Exif.Image.DateTime                          Ascii      20  2015:04:17 19:10:22                                                                                                                                 
Xmp.xmp.ModifyDate                           XmpText    20  2015-04-17T18:10:22Z                                                                                                                                
```
MacOS-X:
```
Exif.Image.DateTime                          Ascii      20  2015:04:18 02:10:22
Xmp.xmp.ModifyDate                           XmpText    20  2015-04-17T18:10:22Z
```

**How to fix this**

This is hard.  Time conversions on v0.26 were performed with src/localtime.c and src/gmtime.h.  Gilles asked us to reinstate src/localtime.c because he was getting linking errors on MinGW/msys2.  I hoped that fix would resolve this.

**Comment by clanmills** [comment]
I'm currently working on the branch 0.27-RC2 and I believe Luis submitted localtime.c to be compiled on MinGW.  test/conversions.sh is OK on the 32bit build and broken on the 64bit build.
```
C:\Users\rmills>msys32
222 rmills@rmillsmm-w7-32:/c/Users/rmills $ cd ~/gnu/github/exiv2/exiv2/build
223 rmills@rmillsmm-w7-32:~/gnu/github/exiv2/exiv2/build $ cd ../test
224 rmills@rmillsmm-w7-32:~/gnu/github/exiv2/exiv2/test $ export EXIV2_BINDIR=$PWD/../build/bin
225 rmills@rmillsmm-w7-32:~/gnu/github/exiv2/exiv2/test $ ./conversions.sh
all testcases passed.
226 rmills@rmillsmm-w7-32:~/gnu/github/exiv2/exiv2/test $ exit
exit

C:\Users\rmills>msys64
501 rmills@rmillsmm-w7-64:/c/Users/rmills $ cd ~/gnu/github/exiv2/exiv2/test
502 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/exiv2/test $ export EXIV2_BINDIR=$PWD/../build/bin
503 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/exiv2/test $ ./conversions.sh
Files /home/rmills/gnu/github/exiv2/exiv2/test/tmp/conversions.out and /home/rmills/gnu/github/exiv2/exiv2/test/data/conversions.out differ
165c165
< Exif.Image.DateTime                          Ascii      20  2015:04:17 19:10:22
---
> Exif.Image.DateTime                          Ascii      20  2015:04:18 02:10:22
504 rmills@rmillsmm-w7-64:~/gnu/github/exiv2/exiv2/test $
```

**Comment by clanmills** [comment]
Eureka!  I've updated the platform on MinGW/msys2 64bit and now it passes the test suite!  I've updated README.md about applying updates to all installed packages on MinGW.

And, I believe we can now get rid of localtime.c.  I do more testing and build with Visual Studio.  I feel confident that this issue will shortly be closed.  Dead.  Dead.  Never come back.

**Comment by clanmills** [comment]
I'm going to close this because the issue has disappeared with upgrading the all installed packages in msys64.  `$ pacman -Syu` This worked on both my machines.  Curiously, a similar upgrade msys32 on both machines was disastrous and exiv2 wouldn't build at all.  I reinstalled msys32 and all the dependencies listed in README.md and all was fine.

I've also deleted `src/localtime.c` and `src/tzfile.h` and commented off the MING conditional compilation of localtime.c in src/CListFiles.txt.  That worked.  It looks as though we can remove that code.

@piponazo and I discussed this on Slack.  We'll leave everything the way it is for v0.27 because we believe Gilles wants this to build for digiKam.  Early in v0.28, this stuff will be removed permanently.

**Comment by clanmills** [comment]
I believe this is still present on MSVC builds.

**Comment by clanmills** [comment]
I'm going to close this again as I don't believe this warrants the effort to debug and fix.  I've tried lots of variants of including/excluding localtime.c in the msvc build and small changes to timegm.h  

I'm going to update team/contrib/buildserver/build.sh to run the test suite on msvc builds.  Those have to be run (as documented in README.txt) using msys64 instead of directly from CMake.

In v0.28, I hope we'll rewrite the time code _AND_ support ctest.

**Comment by clanmills** [comment]
This matter has surfaced again in Exiv2 v0.27.3 when running the test suite on a cross-platform build. https://github.com/Exiv2/exiv2/issues/922#issuecomment-610310126

I've submitted a fix: #1152
